### ISSUE TITLE
Error loading the model

### ISSUE THREAD

**Issue by aralroca** [thekevinscott/ml-classifier#14]
Trying to reproduce the README example:

```js
    const loadImage = (src) => new Promise((resolve, reject) => {
      const image = new Image();
      image.src = src;
      image.crossOrigin = 'Anonymous';
      image.onload = () => resolve(image);
      image.onerror = (err) => reject(err);
    });

    const pretrainedModelURL = 'https://storage.googleapis.com/tfjs-models/tfjs/mobilenet_v1_0.25_224/model.json';

    tf.loadModel(pretrainedModelURL).then(model => {
      const layer = model.getLayer('conv_pw_13_relu');
      return tf.model({
        inputs: [model.inputs[0]],
        outputs: layer.output,
      });
    }).then(pretrainedModel => {
      return tf.loadModel('/model.json').then(model => {
        return loadImage('/trees/tree1.png').then(loadedImage => {
          const image = tf.reshape(tf.fromPixels(loadedImage), [1,224,224,3]);
          const pretrainedModelPrediction = pretrainedModel.predict(image);
          const modelPrediction = model.predict(pretrainedModelPrediction);
          const prediction = modelPrediction.as1D().argMax().dataSync()[0];
          console.log(prediction);
        });
      });
    }).catch(err => {
      console.error('Error', err);
    });
```

I have an error from the line:

```js
const pretrainedModelPrediction = pretrainedModel.predict(image);
```

The error is the next one:

```
tfjs.js:67 Uncaught (in promise) Error: The dtype of the feed (int32) is incompatible with that of the key 'input_1' (float32).
    at new t (tfjs.js:67)
    at assertFeedCompatibility (tfjs.js:67)
    at e.add (tfjs.js:67)
    at new e (tfjs.js:67)
    at tfjs.js:67
    at tfjs.js:49
    at e.scopedRun (tfjs.js:49)
    at e.tidy (tfjs.js:49)
    at e.tidy (tfjs.js:49)
    at s (tfjs.js:67)
```

Any idea about this error? My tfjs version is `0.12.2`.

**Comment by aralroca** [comment]
I uploaded my code into my GitHub if it helps: https://github.com/aralroca/skin-cancer-detection-tfjs/tree/d0d288c84919410dd422a1a19de7b207b6f49000

**Comment by thekevinscott** [comment]
[On this line](https://github.com/aralroca/skin-cancer-detection-tfjs/blob/master/src/SkinDetection.jsx#L20), change ```const image = tf.reshape(tf.fromPixels(img), [1, 224, 224, 3]);``` to be ```const image = tf.reshape(tf.fromPixels(img), [1, 224, 224, 3]).toFloat();```

The message `The dtype of the feed (int32) is incompatible` implies your tensor is of type `int32`, and it needs to be a `float32`. That should make it go through the pipes.

[For more info check out this section.](https://thekevinscott.com/image-classification-with-javascript/#4-translate-the-tensor)

**Comment by aralroca** [comment]
@thekevinscott thank you a lot for the answer. Changing the line you say is fixing this error! 👏

However, I have a question now.

As the docs you shared says, on `[1, 224, 224, 3]`, the first digit is the number of training examples. In my case I have two cases (benign or malignant), should I change it to `[2, 224, 224, 3]`? Because when I do this change, is throwing an error saying that `Error: Size(150528) must match the product of shape 2,224,224,3`. So looks that `1` is correct in my case... But, then the prediction result is always `1` for two cases (benign and malignant)...

For the training process, I used your UI tool, drag & dropping the`train` folder with the two subfolders of images inside.

Thank you a lot again for your time 🙌 , By the way, the UI you created to train the model is so amazing!

**Comment by thekevinscott** [comment]
> As the docs you shared says, on `[1, 224, 224, 3]`, the first digit is the number of training examples. In my case I have two cases (benign or malignant), should I change it to `[2, 224, 224, 3]`? Because when I do this change, is throwing an error saying that `Error: Size(150528) must match the product of shape 2,224,224,3`. So looks that `1` is correct in my case... But, then the prediction result is always `1` for two cases (benign and malignant)...

You'd still want to create the tensor the same way but you then concat them together with `tf.concat` - the tensor you feed to the model should be of shape `[2,224,224,3]` for two images ([though it looks like the latest TF version may just accept a regular array as well](https://stackoverflow.com/questions/58451263/create-a-tensor-of-multiple-images-with-tf-browser-frompixels-with-tensorflow-js)?)

Something like:

```
    const imgs = document.querySelectorAll("img");

    let pixels = tf.reshape(tf.fromPixels(imgs[0]), [1, 224, 224, 3]).toFloat();
    for (let i = 1; i < imgs.length; i++) {
      const img = imgs[i];
      pixels = pixels.concat(tf.reshape(tf.fromPixels(img), [1, 224, 224, 3]).toFloat());
    }
    console.log(pixels.shape);
```

The final `pixels` shape should be `[n, 224, 224, 3]`

> By the way, the UI you created to train the model is so amazing!

Thank you for your kind words, they mean a lot!

**Comment by aralroca** [comment]
@thekevinscott thanks for answering! :) Your feedback is incredible. 

I'm still confusing about the example in the README that I shared in the first comment.

The line 

```js
console.log(prediction);
```

It's always returning `1` for all images. However, during the test with the UI tool, I used the same images and they were correctly predicted, with 88% of evaluation.

After reading a little bit in your blog, I see that in some examples, you replace this:

```js
const image = tf.reshape(tf.fromPixels(loadedImage), [1,224,224,3]);
```

to

```js
 const image = tf.fromPixels(loadedImage).expandDims(0).toFloat().div(tf.scalar(127)).sub(
    tf.scalar(1),
  );
```

What's the difference?

I replaced in my code to try it: https://github.com/aralroca/skin-cancer-detection-tfjs/blob/22d2f16085ecde37aab88968628e7426669bec0e/src/SkinDetection.jsx#L20

And now the prediction is always `0`.

I'm quite lost using the saved model correctly 😅 I will be very grateful if you can provide me some working example different than the README code. Perhaps the README code should be updated?

Anyway, thank you so much. When I will have my saved model working correctly I'm going to recommend your UI tool to everyone! 👏

**Comment by aralroca** [comment]
Doing this:

```js
const pixels = tf.browser.fromPixels(document.querySelector("img"));
const image = tf.reshape(pixels, [1, 224, 224, 3]).toFloat();
const modelPrediction = model.predict(pretrainedModel.predict(image));
const [benign, malignant] = Array.from(modelPrediction.dataSync());
console.log({ benign, malignant })
```

I have the percentages of `benign` and `malignant` from an image. But it no correspond with the tested images using the `ml-classifier-ui` demo, when the evaluation was around 88%, and loading the model in my projects now is failing for a lot of cases, even these tested images... Any idea about what can happen? The images are 224x224 

Thanks 👏

BTW; I'm loading the model doing this:

```js
import * as tf from "@tensorflow/tfjs";
import { useEffect, useState } from "preact/hooks";

const pretrainedModel = {
  url:
    "https://storage.googleapis.com/tfjs-models/tfjs/mobilenet_v1_0.25_224/model.json",
  layer: "conv_pw_13_relu",
};

export default function useLoadSkinDetectionModel() {
  const [state, setState] = useState([]);

  useEffect(() => {
    async function loadModel() {
      const mobilenet = await tf.loadLayersModel(pretrainedModel.url);
      const layer = mobilenet.getLayer(pretrainedModel.layer);
      const pretrained = await tf.model({
        inputs: mobilenet.inputs,
        outputs: layer.output,
      });

      const model = await tf.loadLayersModel(
        "./model/ml-classifier-malignant-benign.json",
      );

      setState([model, pretrained]);
    }
    loadModel();
  }, []);

  return state;
}
```

**Comment by thekevinscott** [comment]
MobileNet expects the incoming tensors to have values between -1 and 1 ([there's a good discussion of why on this Stack Overflow post](https://stackoverflow.com/questions/4674623/why-do-we-have-to-normalize-the-input-for-an-artificial-neural-network)).

Your first question:

> After reading a little bit in your blog, I see that in some examples, you replace this:
> ... 
> What's the difference?

Is the key.

`reshape` in this context is (I think) identical to `expandDims`, but you still need to ```.div(tf.scalar(127)).sub(
    tf.scalar(1),
  );``` the resulting tensor. Otherwise, you're feeding it values between 0 and 255 which it won't know how to interpret.

In your code, make the following modification:

```
const pixels = tf.browser.fromPixels(document.querySelector("img"));
// MODIFY THE FOLLOWING LINE TO DIVIDE AND SUBTRACT
const image = tf.reshape(pixels, [1, 224, 224, 3]).toFloat().div(tf.scalar(127)).sub(
    tf.scalar(1),
  );
const modelPrediction = model.predict(pretrainedModel.predict(image));
const [benign, malignant] = Array.from(modelPrediction.dataSync());
console.log({ benign, malignant })
```

I don't have a great collection of skin images, but I replaced your model with one trained on cats and dogs and it seemed to work with that change.

---

You didn't ask for the below, but I'll take a sec to pontificate in case others land on this issue :)

It can be tricky to diagnose a failing neural net, since problems can fall into any of the three buckets:

1) Problems with the code (usually related to data preprocessing)
2) Problems with how the model was trained
3) Problems with the network's architecture

With something like a web app, a bug often appears as a big red error you can examine. With neural nets, usually it's not as obvious. Neural nets' errors often manifest as the net not predicting correctly, or predicting less accurately than you'd expect. Which makes the bug really hard to diagnose!

In my experience, it's best to start by investigating the first case - your data preprocessing step - to exhaustion. I like to first set up a toy example that I know works (in this case, it'd be something like dogs vs cats) and then unit test the hell out of every step of my data processing pipeline.

Once you've got confidence that the data going in is clean and uncorrupted, it helps narrow down any problems with the second case, how the model was trained. For instance, if the network is able to predict cats and dogs to 90% accuracy, but skin cancer is much lower, you could ask questions like: is my data set imbalanced? Do I have enough data to train? Are my training images representative of the images being used for inference? (Shameless plug: [I talk about data prep more in my book](https://dljsbook.com)!) MobileNet is trained on [ImageNet](http://image-net.org), which is a very diverse collection of images, but I don't think there's many example of skin in the dataset. So you may need more data than with a comparable set of cats vs dogs to see good results. Relatedly, you may find better accuracy by choosing an earlier layer of the original pertained model to start with, since some of the later layers of MobileNet may be irrelevant to the skin cancer use case.

But my point is, if you start asking these questions _before_ you've investigated the data preprocessing step, you could be trying to solve the wrong thing.

(Case 3, neural net architecture issues, I generally don't encounter, since I'm usually using model architectures that others have researched.)

---

Anyways, apologies if you already know all this, I figured it'd might be helpful for others that land on this issue. Let me know if you're still running into issues.

**Comment by aralroca** [comment]
@thekevinscott Thank you a lot for your amazing feedback 👏

I used the `ml-classifier-ui` demo to train the model (just drag&drop).

To be honest I wanted to use the "skin cancer detection" just to learn how to load correctly a model trained with your tool. No for real usage (for the moment). I used [this dataset](https://www.kaggle.com/fanconic/skin-cancer-malignant-vs-benign) from Kaggle as an example. But I seem to have chosen a complicated example. I didn't know how much influence the neural network layer had on transfer learning. I thought that the example layer would be useful for me, but it seems that I should use an earlier one for my case.

Now I've tried the example of [cats and dogs](https://www.kaggle.com/tongpython/cat-and-dog), and as you say, it works perfectly.

I will continue to sweep up a little bit to get to train the example of skin cancer detection with other layers to see what the results are. 

Anyway, now I understand what is happening and I can keep trying.

Is there a physical version of your book? Or just pdf, .mobi? I'd really like to have it on paper. 

I'm going to close the issue.

Thank you very much! I'm going to recommend your tool to everyone!

**Comment by thekevinscott** [comment]
> To be honest I wanted to use the "skin cancer detection" just to learn how to load correctly a model trained with your tool.

Ah, gotcha. Skin cancer detection is certainly more complicated than cats vs dogs, but I wouldn't say it's prohibitively so. The great thing about Kaggle is you can see what's worked for others, and use their results to figure out what kind of accuracy you should be shooting for.

It looks like the top two kernels both use ResNet50 which is a popular architecture. [I found this on npm](https://www.npmjs.com/package/resnet_imagenet), you could also try converting an existing model using [tensorflowjs converter](https://www.tensorflow.org/js/guide/conversion). 

On the other hand, I believe that MobileNet is competitive with ResNet50 (and is tuned to run well in browsers / phones) so it may be a matter of trial-and-error to see which layer provides the sweet spot of accuracy is.  (To change the input layer, you'd want to pass in a custom [`pretrainedModel` with a different layer](https://github.com/thekevinscott/ml-classifier#parameters).)

One thing to keep in mind is that as you go to earlier layers, the training time you'll need will increase. At some point, it won't be feasible to train in the browser. Luckily, you can train in Node! However, you'll need to rewrite the pixel-loading portion to work in Node, as those functions are browser-only.

> Is there a physical version of your book? Or just pdf, .mobi? I'd really like to have it on paper.

Stay tuned! I'm working on it. I agree with you, I prefer paper too :)

> Thank you very much! I'm going to recommend your tool to everyone!

Thank you, and let me know if you run into any other issues!

**Comment by aralroca** [comment]
I wrote about your tool in my blog 😊 https://dev.to/aralroca/classify-any-image-on-the-browser-in-few-steps-pmm Thanks a lot @thekevinscott

**Comment by kevinpelago** [comment]
This is awesome! Great article, and thanks for the shout out!
